PXC 5.7.40 merge part1

VarunNagaraju opened the pull request.

voodam committed:
PS-4759: Percona Server Crashing when the stored procedure called the second time in the same session

https://jira.percona.com/browse/PS-4759

Bug was fixed already in commit f3019ec2978ce9c00fedfbbe6d51120290be28da, just added testcase for it.

i-rinat committed:
Fixed PS-8018 (Parallel replication with blackhole table on replica throws error 1863)

https://jira.percona.com/browse/PS-8018

Problem:

Applying DELETE commands on a replica causes an error that stops
the replication process if the involved table uses BLACKHOLE engine and
is subject to table scans for finding a particular row. In other words,
has no usable indexes.

When applying DELETEs, Partition_helper::ph_delete_row() checks that
the partition ID that was determined by partition info, is exactly the
partition ID from Partition_helper::m_last_part. Latter is set in
Partition_helper::ph_rnd_next() to an ID of the first partition that
reports having the requested row. Since BLACKHOLE engine stores nothing,
it searches nothing. On replicas it happily reports that it has any
requested row. Thus making Partition_helper::ph_rnd_next() believe that
partition 0 is the one that has the row. Partition_helper::m_last_part is
therefore always set to 0 when BLACKHOLE engine is in use. That's fine for
rows in the first partition, but causes HA_ERR_ROW_IN_WRONG_PARTITION for
other partitions.

Fix:

BLACKHOLE engine has no need to track specific rows, so the solution is
to disable the check in Partition_helper::ph_delete_row() by making
ha_blackhole::rpl_lookup_rows() always return false.

handler::rpl_lookup_rows() was introduced as a part of optimizations for
TokuDB and RocksDB, and with that introduced a warning that explicitly
mentions "TokuDB/RocksDB" in its text. Receiving such warning when
BLACKHOLE engine is used would be confusing. So the text of the message is
changed to be engine-agnostic.

bkandasa committed:
Raise version number after cloning 5.6.57

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.39

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

Commit:
Bug#34144951: InnoDB restore position crash when descent scan btree

- This assertion error is only on debug build and no problems
  for normal build. It is from debug code problem.
- btr_cur_optimistic_latch_leaves() get previous block in optimistic,
  and reconfirm page_ids after the current block got latched also.
- But btr_block_get() is intended only for certain use, not for optimistic.
  And it has rigid assertion code.

So, btr_cur_optimistic_latch_leaves() should call buf_page_get_gen()
directly with BUF_GET_POSSIBLY_FREED option, to avoid the rigid assertion.

Change-Id: I05b151d627acf66301ed1c8c081e801af15c1813

Commit:
PS-8287 Move the version info to the variable descriptions (5.7)
modified:   doc/source/diagnostics/innodb_show_status.rst
	modified:   doc/source/management/audit_log_plugin.rst
	modified:   doc/source/management/innodb_kill_idle_trx.rst
	modified:   doc/source/myrocks/variables.rst
	modified:   doc/source/performance/aio_page_requests.rst
	modified:   doc/source/performance/prefix_index_queries_optimization.rst
	modified:   doc/source/scalability/innodb_io.rst
	modified:   doc/source/security/data-at-rest-encryption.rst
	modified:   doc/source/tokudb/tokudb_file_management.rst
	modified:   doc/source/tokudb/tokudb_variables.rst

sreedhars committed:
BUG#34299825 Server 5.7 - Windows msi installation requires Visual Studio 2019 x64 Redistributable
    As we recently upgraded to use newer version of Visualstudio.

Change-Id: I14b8192d6a3f035f5c988641fe700ffa97c8f7e6

Commit:
Bug#32912868: dict_table_x_lock_indexes() can cause deadlock with concurrent purge operations

Confirmed carefully also other possibility with index latch order and implicit latch order.
This bug has the both 2 aspects.

[A] Indexes latch order in same table is [secondary --> primary]
 We should be along with description in sync0type.h.
 The current dict_table_x_lock_indexes() violates the order.
 Better to be fixed.

[B] Even if dict_table_x_lock_indexes() is atomic operation,
 still might cause deadlock with
 "[secondary index page]-->[primary index latch]-->[primary index page]"
 operation. (if dict_table_x_lock_indexes() caller will access to [secondary index page].)
 That seems to be allowed order, and included to purge and rollback paths.
 So, dict_table_x_lock_indexes() should exclude other threads' concurrent accesses.
 - User threads: already excluded by mysqld MDL
 - Purge: ## needs to stop during getting latch ##
 - Rollback after recovery: ## needed to be ended already ##

Fixing [B] is enough to fix this bug.
But also [A] is better to be fixed for future fail-tolerance.

(for 5.7 only)
"TRUNCATE TABLE" path also calls dict_table_x_lock_indexes(), and fixed.

Change-Id: I0673af125aef2f4cc14d8b86a3816c71c5becc8a

Commit:
Bug#33876690 - Server may exit when creating views with a chain
of dependencies

Issue:
The stack might get exhausted while creating a nested view which
in turn references a nested view.

This is because the SELECT nesting level limit of 64 is not
enforced while preparing a view.

Fix:
Check the nesting level while preparing a view and throw an
error if limit exceedes.

Change-Id: I4eef53af614d6c2be337e2cba32cdd27c39487ad

salmankhan123417 committed:
BUG#34159579 Bug 25203933 is still present in 5.7

Description: Whenever the grant option privilege was given to an
user, User will be able to list and change the databases.

Analysis: We had been treating the 'GRANT OPTION' privilege same
as other database related privileges ( e.g. SELECT, INSERT ,
UPDATE, EVENT etc) at all three  global, db and table levels.
Hence, this bug was observed.

Fix: This is a backport from 8.0 to 5.7.
The issue is fixed by redefining the ACL macros such that all the
db related operations are separated from GRANT_ACL. And allowing
<SHOW> and <USE> operations only if db operation privileges are
given.

Change-Id: If32e86d8d87bd0e48ccea690b67eab11901ff915

Commit:
WL #15188 Add wrappers for encryption/decryption algorithms.

Added approved key derivation functions (KDF) pbkdf2_hmac and hkdf.
This enables better key derivations in functions AES_ENCRYPT and AES_DECRYPT.
And this results into better encryption.

rb: 16601
Change-Id: I8dc24b41aaf5e59bd8c1a183a6ab32b0d8c7bd60

Commit:
PS-8348 Release notes and release note tasks for 5.7.39-42

new file:   doc/source/_res/rn/5.7.39-highlights.txt
doc/source/_res/rn/deprecated-5.7.39.txt
new file:   doc/source/_res/rn/ps-mysql-blurb.txt
modified:   doc/source/conf.py
new file:   doc/source/release-notes/Percona-Server-5.7.39-42.rst
modified:   doc/source/release-notes/release-notes_index.rst

GBuella committed:
BUG#34380370 truncate table not free mutex in some case

release mutex on all return paths in truncate_t::truncate

Change-Id: I5aeee262503db191ced7773d2456bd32d7007c53

Commit:
Bug#33866103 : MYSQLPUMP [ERROR] (1356) WITH VIEW CONTAINING FROM
               SUBQUERY - REGRESSION

Description:
Mysqlpump may not work in cases where derived tables are generated.
This is due to privileges not being set for those tables.

Fix:
Backport. Derived tables handled separately and privileges set.

Gerrit#16933

Change-Id: Ia31d479ed2d7aba47993d2e646efb84651ee784a

alina-derkach-oaza committed:
Merge branch '5.7' into ps-5.7-8287

alina-derkach-oaza committed:
Merge pull request #4826 from alina-derkach-oaza/ps-5.7-8287

PS-8287 Move the "version info" to the variable descriptions (5.7)

alina-derkach-oaza committed:
Merge branch '5.7' into ps-5.7-8348

alina-derkach-oaza committed:
Merge pull request #4858 from alina-derkach-oaza/ps-5.7-8348

PS-8348 Release notes and release note tasks for 5.7.39-42

patrickbirch committed:
PXB-8288 Remove replacements from RST files (5.7)

M source/changed_in_57.rst
 M source/development.rst
 M source/faq.rst
 M source/feature_comparison.rst
 M source/index.rst
 M source/index_info_schema_tables.rst
 M source/installation.rst
 M source/percona_xtradb.rst
 M source/upgrading_guide_56_57.rst
 M source/upstream-bug-fixes.rst

patrickbirch committed:
Merge branch '5.7' into ps-8288-5.7

patrickbirch committed:
Merge pull request #4862 from patrickbirch/ps-8288-5.7

PXB-8288 Remove replacements from RST files (5.7)

karthik-kamath committed:
BUG#32893161: BACKPORT 28632725 TO 5.7

When using --log-timestamps=system, ISO8601 timestamps in
logging would always show the timezone's default offset to
UTC, even when DST was active.

Note:
This is a backport of the fix for Bug#28632725 to 5.7.

Change-Id: I488f3468cc320de555cea7bb6f911e9c2f82e225

adivinho committed:
PS-8033 percona-server-shared-compat-57 library doesn't support TLS 1.2

adivinho committed:
Merge pull request #4866 from adivinho/5.7

PS-8033 percona-server-shared-compat-57 library doesn't support TLS 1.2

TharunSai committed:
Bug #34094706 : require_secure_transport causes abrupt server exit when checking
                thd->active_vio

Description:
------------
When require_secure_transport is enabled, the server checks the thd->active_vio
pointer in acl_authenticate(). However, this check is not thread-safe as server
can exit abruptly if the connection is killed before checking active_vio.

Fix:
----
Provided mutex protection in the acl_authenticate() method to avoid this problem
and added a nullptr check as well for thd->active_vio.

Change-Id: I42c284dcaf3c988734010f9faf02f9c023dda90d

kamil-holubicki committed:
Merge pull request #4804 from tempesta-tech/ri-PS-8018-blackhole-engine-replica-partitioning

Fixed PS-8018 (Parallel replication with blackhole table on replica throws error 1863)

Commit:
PS-1098: Fix audit_log_flush behavior when used with auto log rotation

https://jira.percona.com/browse/PS-1098

In case auto log rotation is enabled for audit log and
'SET GLOBAL audit_log_flush=1' is used right after log file is rotated
it leads to just rotated file being reopend. Setting audit_log_flush=1
should do nothing except reopening current audit log file.

The reason of wrong behavior is that do_rotate() method uses data
structures of logger handler to store temporary data while checking
and renaming existing log files. Once it is done with this task logger
handler may be not returned to its correct state in some scenarios.

Rewritten do_rotate() method not to use logger handler as temporary data
storage.

oleksandr-kachan committed:
Merge pull request #4899 from oleksandr-kachan/PS-1098-5.7

PS-1098: Fix audit_log_flush behavior when used with auto log rotation (5.7)

Commit:
Implemented PS-8421 (Merge MySQL 5.7.40) (docs)

https://jira.percona.com/browse/PS-8421

***
Updated man pages from MySQL Server 5.7.40 source tarball.
***
Updated 'scripts/fill_help_tables.sql' from MySQL Server 5.7.40 source tarball.

Commit:
Implemented PS-8421 (Merge MySQL 5.7.40 up to 913af2cd484) (merge with conflicts)

https://jira.percona.com/browse/PS-8421

Commit:
Implemented PS-8421 (Merge MySQL 5.7.40 up to 913af2cd484) (conflicts resolved)

https://jira.percona.com/browse/PS-8421

***
VERSION raised to "5.7.40-43.0".
PERCONA_INNODB_VERSION in univ.i raised to "43.0".
***
Fix ssl_bug70489 test. In some configurations there may be WARN_AES_KEY_SIZE
warning in test result.
***
Remove CPP_DEFAULT_PARAM makro.

voodam committed:
PS-8327: ALTER TABLE ... CHECK PARTITION inside stored procedure causes problem

https://jira.percona.com/browse/PS-8327

Add sp_head::MULTI_RESULTS flag for ALTER TABLE command because it may return a result set in case of CHECK PARTITION.

Also priority of error messages is changed in is_not_allowed_in_function: now sp_head::MULTI_RESULTS has the lowest priority, it makes error message more expectable for ALTER TABLE case.

kamil-holubicki committed:
Merge pull request #4733 from tempesta-tech/am-PS-4759

PS-4759: Percona Server Crashing when the stored procedure called the second time in the same session

percona-ysorokin committed:
Implemented PS-8421 (Merge MySQL 5.7.40 up to 913af2cd484) (ROCKSDB_VALGRIND_RUN)

https://jira.percona.com/browse/PS-8421

'-DROCKSDB_VALGRIND_RUN' preprocessor definition is now also
passed to the 'RocksDB' code when top-level 'WITH_VALGRIND'
CMake option is set to 'ON'.

This allows to eliminate
"==10176== Warning: unimplemented fcntl command: 1036"
Valgrind warnings coming from 'rocksdb/env/io_posix.cc' and
false positives coming from not calling some static object
destructors.

Conditional 'ROCKSDB_UBSAN_RUN' / 'ROCKSDB_VALGRIND_RUN' /
'ROCKSDB_JEMALLOC' RocksDB-specific defines grouped together
in the RocksDB Storage Engine 'CMakeLists.txt' file.

Commit:
Implemented PS-8421 (Merge MySQL 5.7.40 up to 913af2cd484) (valgrind "Invalid read of size 8" in rpl.rpl_parallel_multi_db)

https://jira.percona.com/browse/PS-8421

The "Invalid read of size 8" Valgrind error is observed in
rpl.rpl_parallel_multi_db MTR test results.

The https://github.com/percona/percona-server/commit/bdd12ac67a6cb0e1e295457205e9f640c6e9d0ab
change introduced check to prevent dangerous use of rpl writebatch for
rocksdb. In order to find out if the feature is used properly in
Query_log_event::do_apply_event() we check if handlers for all query
tables support the feature. The issue here is that by the time the check
is performed all involved table objects are already released for later
reuse or complettely closed and memory is freed.

To fix the issue corresponding check was changed to be done at
the point where all query table have just been opened.

Commit:
Implemented PS-8421 (Merge MySQL 5.7.40 up to 913af2cd484) (valgrind "Syscall param pwrite64(buf) points to uninitialised byte")

https://jira.percona.com/browse/PS-8421

Write to uninitialized buffer is used on purpose but it
leads to "Syscall param pwrite64(buf) points to uninitialised byte"
valgrind error in innodb.check_sector_size MTR test.
Use UNIV_MEM_VALID to suppress valgrind error.

Commit:
Implemented PS-8421 (Merge MySQL 5.7.40 up to 913af2cd484) (unit_tests run with valgrind)

https://jira.percona.com/browse/PS-8421

The DartConfiguration.tcl config is missing in buid directory.
The BUILD_TESTING is required to be set for include(CTest) to
generate it properly.

voodam committed:
Merge pull request #4839 from tempesta-tech/am-PS-8327

PS-8327 ALTER TABLE ... CHECK PARTITION inside stored procedure causes problem

oleksandr-kachan committed:
Merge pull request #4909 from oleksandr-kachan/merge-mysql-5.7.40-part1

PS-8421: Merge MySQL 5.7.40 (part1)

VarunNagaraju committed:
Implemented PXC-4049 (Merge PS-5.7.40, Part 1 with conflicts)

https://jira.percona.com/browse/PXC-4049

Merge commit 'fde2cca41b14185d354d2f1ce94dc3ac5b3cbd39' into 5.7.40-merge-part1

Conflicts:
        both modified:   MYSQL_VERSION
        both modified:   config.h.cmake
        deleted by us:   doc/source/changed_in_57.rst
        both modified:   doc/source/conf.py
        deleted by us:   doc/source/development.rst
        deleted by us:   doc/source/diagnostics/innodb_show_status.rst
        both modified:   doc/source/faq.rst
        deleted by us:   doc/source/feature_comparison.rst
        both modified:   doc/source/index.rst
        deleted by us:   doc/source/index_info_schema_tables.rst
        deleted by us:   doc/source/installation.rst
        deleted by us:   doc/source/management/audit_log_plugin.rst
        deleted by us:   doc/source/management/innodb_kill_idle_trx.rst
        deleted by us:   doc/source/myrocks/variables.rst
        deleted by us:   doc/source/percona_xtradb.rst
        deleted by us:   doc/source/performance/prefix_index_queries_optimization.rst
        both modified:   doc/source/release-notes/release-notes_index.rst
        deleted by us:   doc/source/scalability/innodb_io.rst
        deleted by us:   doc/source/security/data-at-rest-encryption.rst
        deleted by us:   doc/source/tokudb/tokudb_file_management.rst
        deleted by us:   doc/source/tokudb/tokudb_variables.rst
        deleted by us:   doc/source/upgrading_guide_56_57.rst
        deleted by us:   doc/source/upstream-bug-fixes.rst
        both modified:   packaging/WiX/mysql_server.wxs.in
        both modified:   sql/log.h

***
Changes in doc/ were ignored and their PXC versions were retained.

VarunNagaraju committed:
Implemented PXC-4049 (Merge PS-5.7.40, Part 1 conflicts resolved)

https://jira.percona.com/browse/PXC-4049

***
MYSQL_VERSION

Bumped PXC_VERSION_EXTRA to 62.

***
config.h.cmake
sql/log.h

e4f7902 backported a fix regarding --log-timestamps=system
from 8.0 to 5.7 in upstream.

***
packaging/WiX/mysql_server.wxs.in

cbf7b8f introduced a restriction that a 5.7 Windows msi
installation requires Visual Studio 2019 x64 Redistributable.

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1709) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1709) before we can accept your contribution.<br/>**6** out of **15** committers have signed the CLA.<br/><br/>:white_check_mark: alina-derkach-oaza<br/>:white_check_mark: patrickbirch<br/>:white_check_mark: adivinho<br/>:white_check_mark: kamil-holubicki<br/>:white_check_mark: VarunNagaraju<br/>:white_check_mark: voodam<br/>:x: bkandasa<br/>:x: mysql-builder@oracle.com<br/>:x: sreedhars<br/>:x: i-rinat<br/>:x: karthik-kamath<br/>:x: salmankhan123417<br/>:x: percona-ysorokin<br/>:x: oleksandr-kachan<br/>:x: GBuella<br/><hr/>**mysql-builder@oracle.com** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtradb-cluster?pullRequest=1709) it.</sub>

VarunNagaraju commented:
Jenkins test results of the part 1 merge: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/359/#showFailuresLink
Two test failures: main.merge and main.merge_mmap.

kamil-holubicki commented on MYSQL_VERSION:
```diff
@@ -1,13 +1,5 @@
 MYSQL_VERSION_PATCH=40
 MYSQL_VERSION_EXTRA=-43
->>>>>>> fde2cca41b14185d354d2f1ce94dc3ac5b3cbd39
+PXC_VERSION_EXTRA=62
```
This should be bumped to next even number, so 63

kamil-holubicki commented on packaging/WiX/mysql_server.wxs.in:
```diff
@@ -76,44 +76,6 @@
-        </Condition>
-      <?endif?>
-=======
       <?if "@Platform@"="x64"?>
```
Interesting, that we had a conflict in this file... :)

kamil-holubicki commented on sql/log.h:
```diff
@@ -1004,15 +1004,12 @@ bool log_syslog_find_facility(char *f, SYSLOG_FACILITY *rsf);
 #endif /* WITH_WSREP */
-||||||| b0a7dc2da2e
-=======
 /* 26 for regular timestamp, plus 7 (".123456") when using micro-seconds */
```
My suggestion is to move the new block before #ifdef WITH_WSREP.
This way we will have WSREP-related things at the end of the file, not in the middle

VarunNagaraju replied:
Done.

kamil-holubicki commented on build-ps/percona-server.spec:
```diff
@@ -91,8 +91,8 @@
 %if 0%{?rhel} > 6
-%global compatver             5.6.28
-%global percona_compatver     76.1
+%global compatver             5.6.51
```
content of build-ps directory should not be changed

VarunNagaraju replied:
Done.

VarunNagaraju committed:
Implemented PXC-4049 (Merge PS-5.7.40, Fixed minor issues)

https://jira.percona.com/browse/PXC-4049

* Bumped the PXC_VERSION_EXTRA to an odd number(63).
* Reverted the changes in build-ps/percona-server.spec to
  retain the contents of build-ps directory.
* Moved the WSREP specific #ifdef to the bottom of the file.
* Removes DWITH_ROCKSDB and DWITH_TOKUDB compilation options
  from azure pipelines used for github checks.
* Removes macOS platfroms from azure pipelines since they're
  not supported yet for PXC.
* Disables MyRocksDB and TokuDB completely from compiling.
* Added DWITH_WSREP=ON to azure pipelines to fix compilation
  failure in github checks.
* When WITH_KEYRING_VAULT_TEST is enabled, it needs some wsrep_* files
  which use a header from wsrep/src. So, the path is updated in the
  respective CMakeLists.txt.

kamil-holubicki reviewed (approved):
lgtm

VarunNagaraju commented:
Jenkins run with PXB 2.4.27: https://pxc.cd.percona.com/job/pxc-5.7-param/363/
Test failures: 
rpl.rpl_mts_gtids_restart_slave_io_lost_trx_with_master_delay : Seems like a flaky test. Verified.
